### PR TITLE
Borevent snapshot validation

### PR DESCRIPTION
**Pull request by mh0lt**
This adds integrity checking for borevents in the following places

1. Stage Bor Heimdall - check that the event occurs withing the expected time window
2. Dump Events to snapshots - check that the event ids are continuous in and between blocks (will add a time window tests)
3. Index event snapshots -   check that the event ids are continuous in and between blocks

It also adds an external integrity checker which runs on snapshots checking for continuity and timeliness of events.  This can be called using the following command:
 `erigon snapshots integrity --datadir=~/snapshots/bor-mainnet-patch-1 --from=45500000` 
(--to specifies an end block)

This also now fixes the long running issue with bor events causing unexpected fails in executions:

the problem event validation uncovered was a follows:

The **kv.BorEventNums** mapping table currently keeps the mapping first event id->block.  The code which produces bor-event snapshots to determine which events to put into the snapshot.

however if no additional blocks have events by the time the block is stored in the snapshot, the snapshot creation code does not know which events to include - so drops them.

This causes problems in two places:

  * RPC queries & re-execution from snapshots can't find these dropped events
  * Depending on purge timing these events may erroneously get inserted into future blocks
  
 The code change in this PR fixes that bug.
 
It has been tested by running:

```
erigon --datadir=~/chains/e3/amoy --chain=amoy --bor.heimdall=https://heimdall-api-amoy.polygon.technology --db.writemap=false --txpool.disable --no-downloader --bor.milestone=false
```

with validation in place and the confimed by running the following:

```
erigon snapshots rm-all-state-snapshots --datadir=~/chains/e3/amoy
rm ~/chains/e3/amoy/chaindata/
erigon --datadir=~/chains/e3/amoy --chain=amoy --bor-heimdall=https://heimdall-api-amoy.polygon.technology --db.writemap=false --no-downloader --bor.milestone=false
```
To recreate the chain from snapshots.

It has also been checked with:

```
erigon snapshots integrity --datadir=~/chains/e3/amoy --check=NoBorEventGaps --failFast=true"
```

### CODE REVIEW

File: polygon/bor/bor.go [hunk 1]
```diff
@@ -233,6 +233,35 @@ type ValidateHeaderTimeSignerSuccessionNumber interface {
 	GetSignerSuccessionNumber(signer libcommon.Address, number uint64) (int, error)
 }
 
+func CalculateEventWIndow(ctx context.Context, config *borcfg.BorConfig, header *types.Header, tx kv.Getter, headerReader services.HeaderReader) (from time.Time, to time.Time, err error) {
```
Comment (taratorio) [inline polygon/bor/bor.go:236]: typo `CalculateEventWIndow` - `CalculateEventWindow`

File: polygon/bor/genesis_contracts.go [hunk 2]
```diff
@@ -26,10 +28,21 @@ var (
 func GenesisContractValidatorSetABI() abi.ABI {
 	return validatorSetABI
 }
+
 func GenesisContractStateReceiverABI() abi.ABI {
 	return stateReceiverABI
 }
 
+var methodId []byte = stateReceiverABI.Methods["commitState"].ID
```
Comment (taratorio) [inline polygon/bor/genesis_contracts.go:36]: `[]byte` type can be omitted

File: eth/stagedsync/bor_heimdall_shared.go [hunk 3]
```diff
@@ -220,19 +217,21 @@ func fetchAndWriteHeimdallStateSyncEvents(
 			continue
 		}
 
-		if lastStateSyncEventID+1 != eventRecord.ID || eventRecord.ChainID != chainID || !eventRecord.Time.Before(to) {
+		if lastStateSyncEventID+1 != eventRecord.ID || eventRecord.ChainID != chainID ||
+			!(eventRecord.Time.After(from) && eventRecord.Time.Before(to)) {
```
Comment (taratorio) [inline eth/stagedsync/bor_heimdall_shared.go:221]: I'd suggest to add a new unit test for this error path - it should be fairly easy to add.

You can have a look at `TestBorHeimdallForwardPersistsStateSyncEvents` and `stagedsynctest.Harness.mockHeimdallClient`. For this you will need to pass in a new function to `h.heimdallClient.EXPECT().FetchStateSyncEvents(...).DoAndReturn()`. A clean way to do this may be to add an optional `MockFetchStateSyncEventsCallback` to the `HarnessCfg` and check if it is present use that inside the DoAndReturn otherwise use the current one.

File: eth/integrity/snap_blocks_read.go [hunk 4]
```diff
@@ -10,20 +10,25 @@ import (
...
 	defer logEvery.Stop()
 
 	maxBlockNum := blockReader.Snapshots().SegmentsMax()
-	for i := uint64(0); i < maxBlockNum; i += 10_000 {
+
+	if to != 0 && maxBlockNum > to {
+		maxBlockNum = 2
+	}
+
+	for i := uint64(from); i < maxBlockNum; i += 10_000 {
```
Comment (taratorio) [inline eth/integrity/snap_blocks_read.go:24]: error message here - unnecessary conversion

File: eth/integrity/snap_blocks_read.go [hunk 5]
```diff
@@ -10,20 +10,25 @@ import (
...
+func SnapBlocksRead(ctx context.Context, db kv.RoDB, blockReader services.FullBlockReader, from, to uint64, failFast bool) error {
 	defer log.Info("[integrity] SnapBlocksRead: done")
 	logEvery := time.NewTicker(10 * time.Second)
 	defer logEvery.Stop()
 
 	maxBlockNum := blockReader.Snapshots().SegmentsMax()
-	for i := uint64(0); i < maxBlockNum; i += 10_000 {
+
+	if to != 0 && maxBlockNum > to {
+		maxBlockNum = 2
```
Comment (taratorio) [inline eth/integrity/snap_blocks_read.go:21]: out of curiosity - why is this necessary? is it worth leaving a comment?

File: eth/stagedsync/bor_heimdall_shared.go (hunk 3 above)
Comment (mh0lt) [reply eth/stagedsync/bor_heimdall_shared.go:221]: ok I'll look at this tomorrow.

File: eth/integrity/snap_blocks_read.go (hunk 5 above)
Comment (mh0lt) [reply eth/integrity/snap_blocks_read.go:21]: hmm - should say: maxBlockNum = to

File: eth/stagedsync/bor_heimdall_shared.go [hunk 6]
```diff
@@ -153,21 +133,37 @@ func fetchAndWriteHeimdallStateSyncEvents(
...
+	prevHeader, err := blockReader.HeaderByNumber(ctx, tx, blockNum-config.CalculateSprintLength(blockNum))
+
+	if err != nil {
+		return lastStateSyncEventID, 0, time.Since(fetchStart), err
+	}
+
 	if config.IsIndore(blockNum) {
 		stateSyncDelay := config.CalculateStateSyncDelay(blockNum)
 		to = time.Unix(int64(header.Time-stateSyncDelay), 0)
+		from = time.Unix(int64(prevHeader.Time-stateSyncDelay), 0)
```
Comment (taratorio) [inline eth/stagedsync/bor_heimdall_shared.go:157]: can the new func `CalculateEventWIndow` you have added be used here as well instead of this if else?

**Comment by AskAlexSharov** [comment]
did try this "integrity" check, see:
```
EROR[02-20|04:59:03.985] [integrity] NoGapsInBorEvents: invalid event time block=25196544 event=1716410 time=2022-02-21T14:37:02+0000 expected="2022-02-21 14:38:38 +0000 UTC-2022-02-21 14:38:38 +0000 UTC"
```
for all events

**Comment by mh0lt** [comment]
> did try this "integrity" check, see:
> 
> ```
> EROR[02-20|04:59:03.985] [integrity] NoGapsInBorEvents: invalid event time block=25196544 event=1716410 time=2022-02-21T14:37:02+0000 expected="2022-02-21 14:38:38 +0000 UTC-2022-02-21 14:38:38 +0000 UTC"
> ```
> 
> for all events

I'm only just testing this now.  I need to validate that the check is actually correctly configured - which I'm talking through with Polygon.  As there is no spec for the start of a window its going to take some adjustment to get right.

**Review by AskAlexSharov — APPROVED** [review]

File: polygon/bor/borabi/abi.go [hunk 7]
```diff
@@ -51,3 +55,13 @@ func ValidatorSetContractABI() abi.ABI {
 func StateReceiverContractABI() abi.ABI {
 	return stateReceiverABI
 }
+
+var methodId []byte = stateReceiverABI.Methods["commitState"].ID
+
+func EventTime(encodedEvent rlp.RawValue) time.Time {
```
Comment (taratorio) [inline polygon/bor/borabi/abi.go:61]: this func is not used (duplicate with the one in the heimdall package) and can be removed

File: eth/integrity/no_gaps_bor_events.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+					return err
+				}
+
+				bodyProgress, err := stages.GetStageProgress(tx, stages.Bodies)
+
+				if err != nil {
+					return err
+				}
+
+				fmt.Println("LAST Event", lastEventId, "BH", borHeimdallProgress, "B", bodyProgress)
```
Comment (taratorio) [inline eth/integrity/no_gaps_bor_events.go:110]: looks like leftover print from debugging? should we remove and/or use a logger?

File: polygon/heimdall/event_fetch_test.go [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/erigontech/erigon-lib/log/v3"
+)
+
+func TestOver50EventBlockFetch(t *testing.T) {
```
Comment (taratorio) [inline polygon/heimdall/event_fetch_test.go:11]: this test should maybe be removed from main since it is calling an external service `https://heimdall-api.polygon.technology/` and may fail our CI

File: polygon/heimdall/event_record.go [hunk 10]
```diff
@@ -131,3 +131,29 @@ type StateSyncEventResponse struct {
 	Height string              `json:"height"`
 	Result EventRecordWithTime `json:"result"`
 }
+
+var methodId []byte = borabi.StateReceiverContractABI().Methods["commitState"].ID
```
Comment (taratorio) [inline polygon/heimdall/event_record.go:135]: `[]byte` type can be omitted in the declaration

File: eth/integrity/no_gaps_bor_events.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+
+				if bodyProgress > borHeimdallProgress {
```
Comment (taratorio) [inline eth/integrity/no_gaps_bor_events.go:112]: this if and for loop don't seem to do anything?
probably the whole `if db != nil` if branch above on line 89 can be removed?

File: eth/integrity/snap_blocks_read.go (hunk 5 above)
Comment (taratorio) [reply eth/integrity/snap_blocks_read.go:39]: @mh0lt this typo still needs fixing

**Review by taratorio — APPROVED** [review]
